manual/src/tutorials/effectexamples.etex

```diff
@@ -0,0 +1,666 @@
+
+let _ = run (fun _ ->
+  fork (fun _ ->
+    printf "[t1] Sending 0\n";
+    printf "[t1] received %d\n" (xchg 0));
+  fork (fun _ ->
+    printf "[t2] Sending 1\n";
+    printf "[t2] received %d\n" (xchg 1)))
+\end{caml_example}
+
+Observe that the messages from the two tasks are interleaved. Notice also that
+the snippet above makes no reference to the effect handlers, and is in
+direct-style (no monadic operations). This example illustrates that, with
+effect handlers, the user code in a concurrent program can remain in simple
+direct-style, and the use of effect handlers can be fully contained within the
+concurrency library implementation.
+
+\subsection{s:effects-discontinue}{Resuming with an exception}
+
+In addition to resuming a continuation with a value, effect handlers also
+permit resuming by raising an effect at the point of perform. This is done with
+the help of "discontinue" primitive. Discontinue helps ensure that
+resource-safe code in OCaml behaves correctly.
+
+For example, consider the dequeue operation in the previous example reproduced
+below:
+
+\begin{caml}
+\begin{camlinput}
+let dequeue () =
+  if Queue.is_empty run_q then () (* done *)
+  else (Queue.pop run_q) ()
+\end{camlinput}
+\end{caml}
+
+If the scheduler queue is empty, dequeue considers the scheduler to be complete
+and returns to the caller. However, there may still be a pending task waiting
+to exchange a value, which remains blocked forever!. If the blocked task holds
+onto resources, such resources are leaked. For example, consider the following
+task:
+
+\begin{caml_example*}{verbatim}
+let leaky_task () =
+  fork (fun _ ->
+    let oc = open_out "secret.txt" in
+    Fun.protect ~finally:(fun _ -> close_out oc) (fun _ ->
+      output_value oc (xchg 0)))
+\end{caml_example*}
+
+The task writes the received message to the file "secret.txt". It uses
+"Fun.protect" to ensure that the output channel "oc" is closed on both normal
+and exceptional return cases. If this task does not have a matching exchange,
+then the task will remain blocked forever, and leaks the output channel "oc".
+
+We can use "discontinue" to ensure that the blocked task gets a chance to free
+up resources:
+
+\begin{caml}
+\begin{camlinput}
+exception Improper_synchronization
+
+let dequeue () =
+  if Queue.is_empty run_q then begin
+    match !exchanger with
+    | None -> () (* done *)
+    | Some (n, k) ->
+        exchanger := None;
+        discontinue k Improper_synchronization
+  end else (Queue.pop run_q) ()
+\end{camlinput}
+\end{caml}
+
+When the scheduler queue is empty and there is a blocked exchanger thread, the
+dequeue function discontinues the blocked thread with
+"Improper_synchronization" exception. This exception is raised at the blocked
+"xchg" function call, which makes the "finally" block to be run and close the
+output channel "oc".
+
+\section{s:effects-generator}{Generators}
+
+Generators are a language mechanism that permits non-continuous traversal of a
```

Comment:
IMHO, "non-continuous" is unclear; one might take it to mean that the traversal can skip parts of the data structure, or something like that.

Furthermore, I would suggest avoiding the words "generator" and "iterator", which are heavily overloaded; or, if you must use them, defining them carefully. Using the word "iterator" to designate an OCaml `iter` function definitely seems non-standard to me; I would suggest sticking with "an `iter` function" instead. Regarding functions of type `unit -> 'a option`, I would advise against the use of the word "generator". In the documentation of `Seq` in OCaml 4.14, we have used "dispenser" as a neutral word for this concept.

Reply:
This also reminds me that it is preferable to first convert from an `iter` function into a side-effect free sequence of type `Seq.t`. Then, if a stateful dispenser is desired, the function `Seq.to_dispenser` (new in 4.14) can be used. This leads to more elegant code and a separation of concerns.

Reply:
See [Figure 8 of my paper with Paulo de Vilhena](http://cambium.inria.fr/~fpottier/publis/de-vilhena-pottier-sleh.pdf). In the absence of mutable state, the code is simpler; obviously, there is no mutable reference; furthermore, there is no need for a recursively-defined `step` function.

Furthermore, by removing mutable state, we no longer need to choose between the words "generator" and "dispenser": we can work with "sequences" in the sense of OCaml's `Seq` module.

Reply:
I played around a little based on the example from the paper. Avoiding mutable state is a win. However, the fact that the continuations are one-shot will mean that the results of the inversion will need to be a dispenser IINM. The following snippet shows a negative example:

```ocaml
open Effect
open Effect.Deep

let invert (type a) (iter : (a -> unit) -> unit) : a Seq.t =
  let module M = struct
    type _ Effect.t += Yield : a -> unit t
  end in
  let yield v = perform (M.Yield v) in
  fun () -> match_with iter yield
  { retc = (fun _ -> Seq.Nil);
    exnc = raise;
    effc = fun (type b) (eff : b Effect.t) ->
      match eff with
      | M.Yield v -> Some (fun (k: (b,_) continuation) ->
          Seq.Cons (v, continue k))
      | _ -> None }

(* List *)
let lseq = invert (Fun.flip List.iter [1;2;3])
let next = Seq.to_dispenser lseq;;
next ();;
next ();;
next ();;
next ();;
(* Works fine so far. But our continuations are one shot.. *)

let v,ls = Option.get (Seq.uncons lseq)
let _ = Seq.uncons ls
let _ = Seq.uncons ls (* blows up with "Fatal error: exception Continuation_already_taken" *)
```

I'll modify the example to make it a dispenser. I'll add a reference to the `Seq` module documentation when I introduce the dispenser.

Reply:
I shall rewrite this section based on the following code:

```ocaml
open Effect                                                                         
open Effect.Deep                                                                    
                                                                                    
type 'a dispenser = unit -> 'a option                                               
                                                                                    
let invert (type a) (iter : (a -> unit) -> unit) : a dispenser =                    
  let module M = struct                                                             
    type _ Effect.t += Yield : a -> unit t                                          
  end in                                                                            
  let yield v = perform (M.Yield v) in                                              
  let rec step = ref (fun () ->                                                     
    match_with iter yield                                                           
    { retc = (fun _ -> None);                                                       
      exnc = raise;                                                                 
      effc = fun (type b) (eff : b Effect.t) ->                                     
        match eff with                                                              
        | M.Yield v -> Some (fun (k: (b,_) continuation) ->                         
            step := (fun () -> continue k ());                                      
            Some v)                                                                 
        | _ -> None })                                                              
  in                                                                                
  fun () -> !step ()                                                                
                                                                                    
let next = invert (Fun.flip List.iter [1;2;3]);;                                    
next ();;                                                                           
next ();;                                                                           
next ();;                                                                           
next ()
```

I'll avoid the mention of generators, and use dispenser. What do you think?

Reply:
I don't think that the fact that continuations are one-shot is a problem. In spite of this, the function `invert` can produce a sequence (of type `'a Seq.t`), but one must document the fact that this sequence is ephemeral (it can be consumed only once).

Turning this ephemeral sequence into a dispenser can be done by applying `Seq.to_dispenser`, and this results in a safe API (one that cannot be misused), which is good. But I see no fundamental reason to merge the code of `invert` and the code of `Seq.to_dispenser` into a single function; this confuses matters and is not beneficial, I would argue.

Reply:
> but one must document the fact that this sequence is ephemeral (it can be consumed only once)

This sounds good to me. Let me work on this next.

Reply:
It does sound like a good idea.

> Its type could be (('a -> unit) -> unit) -> 'a Seq.t

That's correct. 

> A more neutral name might be Seq.construct or Seq.build or something along these lines.

Given that there are several functions to "construct" sequences in the `Seq` module now (`init`, `unfold`, `repeat`), naming this function with a neutral name like `contruct` or `build` doesn't seem like the right thing. `with_yield` seems to have a good balance. 

> drop method that would allow us to discontinue the captured continuation

Indeed. I would imagine that it is very common not to fully consume a sequence, in which case, `drop` is necessary.

Reply:
@fpottier I've rewritten the section on generators with the `invert` function that converts an `iter` function to a sequence. I'm a little bit unsatisfied with the last paragraph in the section about linearity

> It is important to note that the sequence returned by the invert function is ephemeral (as defined by the [Seq](https://github.com/ocaml/ocaml/api/Seq.html) module). In particular, this sequence must be consumed exactly once since our continuations are linear (see section ‍[9.5.4](https://github.com/ocaml/ocaml/pull/11093#s%3Aeffects-linearity)).

We need a `Seq.drop` to not leak continuations. I plan to address this after this PR gets merged.

Reply:
Hi @kayceesrk, I will read your new text ASAP. I don't see how one could implement `Seq.drop`: the type definition for sequences (which we cannot change) does not seem to allow it. (It does allow an inefficient implementation that repeatedly asks for elements until the sequence is exhausted; is this what you have in mind?)

Reply:
Indeed. You are right about the challenge of implementing `Seq.drop` without changing the interface. The naive solution won't work for infinite sequences:

```ocaml
let rec nats i f = f i; nats (i+1) f
let s = invert (nats 0)
```

Reply:
I wonder if this implies that we should *not* provide `Seq.with_yield`, because it will encourage people to write code where there is a risk of leaking a continuation.

Instead of a function that builds a sequence of type `'a Seq.t`, should we propose a function that builds a sequence of a richer type, one that allows either asking for the next element or stopping? The upside is that this provides a mechanism for avoiding leaks of continuations. The downside is that we depart from the `Seq` API.

Reply:
Unfortunately, it does seem so. I'm discussing with @ctk21 whether we can free fiber stacks for unresumed continuations. But it is a partial solution since we're not aiming to `discontinue` unresumed continuations. For example,

```ocaml
let read_lines file f =                                                             
  let fd = open_in file in                                                          
  let rec loop () = f (input_line fd); loop () in                                   
  Fun.protect ~finally:(fun () -> close_in fd) loop
  
let s = invert (read_lines "input.txt")
```
will leak `fd` even if the stack associated with the fiber was freed. 

If we had a `drop` function for the sequence, then `drop` could `discontinue` the continuation, which would close `fd`. It feels like ephemeral sequences should have an associated `drop` or `close` functionality especially since they may close over resources.

Reply:
For this tutorial, I'm considering whether to change the last paragraph to 

> It is important to note that the sequence returned by the `invert` function is ephemeral (as defined by the [Seq](https://github.com/ocaml/ocaml/api/Seq.html) module). In particular, all the elements in the sequence must be consumed exactly once to ensure that the captured continuation is linearly used (see section ‍9.5.4).

Would that be sufficient to make forward progress on this issue?

Reply:
Yes, these three lines are sufficient for the tutorial. I would suggest clarifying that "ephemeral" implies that the sequence must be used at most once, and (separately) adding that the sequence must be fully consumed (i.e., used at least once) so as to ensure that the captured continuation is used linearly.

Reply:
Done.

Reply:
@fpottier: a pet peeve of mine is that `Seq` should provide information at the end of the sequence:

```
type ('a, 'b) seq = unit -> ('a, 'b) cell
and ('a, 'b) cell = Cons of 'a * ('a, 'b) seq | End of 'b
```

(In particular, this gives a nice, more natural type to `unfold`.)

Maybe this can be combined with the early-stopping idea by using

```
type ('a, 'b) seq = { next : unit -> ('a, 'b) seq; drop : unit -> 'b }
```